Week2/Homework/MySQL exercises/aggreFunctions.js

```diff
@@ -0,0 +1,86 @@
+const util = require('util');
+const mysql = require('mysql');
+const fs=require('fs');
+const config=JSON.parse(fs.readFileSync("config_secret.json"))
+
+const connection = mysql.createConnection({
+  host: config.host,
+  user: config.user,
+  password: config.password,
+  database: config.database
+});
+
+const execQuery = util.promisify(connection.query.bind(connection));
+
+async function seedDatabase() {
+ 
+    const  researchPapers_and_Thier_authors = `
+    SELECT R.paper_id,paper_title,count(A.author_id) FROM research_papers R 
+    LEFT JOIN author_research_papers A
+    ON A.paper_id = R.paper_id
+    group by A.paper_id; ` ;
+
+    const sumOf_female_researcher = `
+    SELECT count(*) AS female_researcher FROM authors A
+    JOIN author_research_papers R
+    ON A.author_no = R.author_id
+    WHERE gender='f' ;`;
+
+    const averageOf_H_index_perUnive= `
+    SELECT university,AVG(h_index)
+    FROM authors
+    GROUP BY university;
+    `
+    const sumOf_researchPapers_perUnive= `
+    SELECT university,count(R.paper_id) AS num_research_papers FROM authors A JOIN author_research_papers R
+    ON A.author_no = R.author_id
+    GROUP BY university;
+    `
+    const minAndMax_Of_H_index_perUnive= `
+    SELECT university,MAX(h_index),MIN(h_index)
+    FROM authors
+    GROUP BY university;
+    `
+
+    
+  connection.connect();
+
+  try {
+     execQuery( researchPapers_and_Thier_authors,(err ,results)=>{
+      if(err) throw err;
+      console.log('research papers and the number of authors are:',results);
+
+    });
```

Comment:
tip: You can make use of the Promise that execQuery will return and `catch` errors and thus remove the `try` statement.

```suggestion
    execQuery(researchPapers_and_Thier_authors)
      .catch(err => {throw err})
      .then(results => {
          console.log('research papers and the number of authors are:', results);
      });
```

Reply:
didnot work as you suggest and I got this error .. so I came back to the try block ..

this is my code and below the error : 
execQuery(researchPapers_and_Thier_authors)
    .catch((err) => {
      throw err;
    })
    .then((results) => {
      console.log('research papers and the number of authors are:', results);
    });
  execQuery(sumOf_female_researcher)
    .catch((err) => {
      throw err;
    })
    .then((results) => {
      console.log(
        'the Sum of the research papers published by all female authors is:',
        results
      );
    });
  execQuery(averageOf_H_index_perUnive)
    .catch((err) => {
      throw err;
    })
    .then((results) => {
      console.log(
        'Average of the h-index of all authors per university is:',
        results
      );
    });
  execQuery(sumOf_researchPapers_perUnive)
    .catch((err) => {
      throw err;
    })
    .then((results) => {
      console.log(
        'Sum of the research papers of the authors per university is:',
        results
      );
    });
  execQuery(minAndMax_Of_H_index_perUnive)
    .catch((err) => {
      throw err;
    })
    .then((results) => {
      console.log(
        'Minimum and maximum of the h-index of all authors per university is:',
        results
      );
    });

  connection.end();
}
connection.end();

seedDatabase();


node:events:353
      throw er; // Unhandled 'error' event
      ^

Error: Cannot enqueue Handshake after invoking quit.
    at Protocol._validateEnqueue (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:215:16)
    at Protocol._enqueue (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:138:13)
    at Protocol.handshake (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:51:23)
    at Connection.connect (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\Connection.js:116:18)
    at seedDatabase (C:\Users\Thaer\Desktop\DB_Homework\databases\Week2\Homework\MySQL exercises\aggreFunctions.js:45:14)
    at Object.<anonymous> (C:\Users\Thaer\Desktop\DB_Homework\databases\Week2\Homework\MySQL exercises\aggreFunctions.js:99:1)
←[90m    at Module._compile (node:internal/modules/cjs/loader:1108:14)←[39m
←[90m    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)←[39m
←[90m    at Module.load (node:internal/modules/cjs/loader:973:32)←[39m
←[90m    at Function.Module._load (node:internal/modules/cjs/loader:813:14)←[39m
Emitted 'error' event on Connection instance at:
    at Connection._handleProtocolError (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\Connection.js:423:8)
←[90m    at Protocol.emit (node:events:376:20)←[39m
    at Protocol._delegateError (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:398:10)
    at Handshake.<anonymous> (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:232:10)
←[90m    at Handshake.emit (node:events:376:20)←[39m
    at Handshake.Sequence.end (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\sequences\Sequence.js:78:12)
    at C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:236:14
←[90m    at processTicksAndRejections (node:internal/process/task_queues:75:11)←[39m {
  code: ←[32m'PROTOCOL_ENQUEUE_AFTER_QUIT'←[39m,
  fatal: ←[33mfalse←[39m
}

Reply:
I've edited your reply using the text editor on Github by adding a code block and a quote block.
That makes it way easier to read your reply.

And I'm surprised by your stance "did not work as suggested... so I came back to try block".
Do you think that I gave you a faulty suggestion?

Reply:
I think your code is looking fine on the `.catch().then()` part and I'll help you understand your error.
I'm usually scanning the whole error to see where I find a line that talks about the file I'm working in.
That on line 6 and 7 in this case so when I read the first 7 lines in reverse you can see this:

> at Object. (C:\Users\Thaer\Desktop\DB_Homework\databases\Week2\Homework\MySQL exercises\aggreFunctions.js:99:1)

That's the line where `seedDatabase()` is.

> at seedDatabase (C:\Users\Thaer\Desktop\DB_Homework\databases\Week2\Homework\MySQL exercises\aggreFunctions.js:45:14)

confirmed by the next line. which starts the function `seedDatabase` and has an error at line 45
That's `connection.connect();`

> at Protocol._validateEnqueue (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:215:16)
at Protocol._enqueue (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:138:13)
at Protocol.handshake (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\protocol\Protocol.js:51:23)
at Connection.connect (C:\Users\Thaer\node_modules\←[4mmysql←[24m\lib\Connection.js:116:18)

And you can see that `connection.connect` is being run in `C:\Users\Thaer\node_modules\mysql\` and the error is triggered there.

>  Error: Cannot enqueue Handshake after invoking quit.

Last but not least is the actual error message complaining about something 'invoking quite' before the handshake even happened.

If you look at the end of your snippet you can see what went wrong.
```js
  connection.end();
}
connection.end();

seedDatabase();
```
You did a `connection.end()` right before you call `seedDatabase()`

Reply:
Supper 👍  ... I am grateful for your explanation ..." Do you think that I gave you a faulty suggestion?" no If course not .. sorry for my expression .. my language might seem rude ..I tried to google the error but I lost :( ... thanks again